### PR TITLE
Remove deprecated org.gradle.util.DistributionLocator usage

### LINKED ISSUE #5494: [BUG] Gradle task `:wrapper` fails with org.gradle.util.DistributionLocator deprecated
*NOT URGENT, but should be looked into at some point*

**Describe the bug**
[Wrapper task](https://github.com/opensearch-project/OpenSearch/blob/main/build.gradle#L532) to update gradle version fails with `The org.gradle.util.DistributionLocator type has been deprecated`.

```
> Task :wrapper
The org.gradle.util.DistributionLocator type has been deprecated. This is scheduled to be removed in Gradle 9.0. Consult the upgrading guide for further information: https://docs.gradle.org/7.6/userguide/upgrading_version_7.html#org_gradle_util_reports_deprecations
        at build_4d0q822uef9ladyhg6p6yk3ft$_run_closure17$_closure85.doCall(.../OpenSearch/build.gradle:535)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)
Added checksum to wrapper properties
Updated minimum Gradle Version

FAILURE: Build failed with an exception.

* Where:
Build file '.../OpenSearch/build.gradle' line: 535

* What went wrong:
Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0
```

**To Reproduce**
Run `./gradlew  wrapper --gradle-version 7.6 --distribution-type all`

**Expected behavior**
Task should be successful.

**Host/Environment (please complete the following information):**
 - OS: `main` as of commit: ce25dec

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

The [org.gradle.util.DistributionLocator](https://docs.gradle.org/current/javadoc/org/gradle/util/DistributionLocator.html) class is deprecated and will be removed in Gradle 9.0.

This PR replaces that usage in `build.gradle` with `wrapper.getDistributionUrl()` which [implements the identical logic internally](https://github.com/gradle/gradle/blob/5aa17c00b3884c66e0ce0f25af5ca749821bbd1f/subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java#L375-L383) (and is not deprecated).

Also added a new line to the min version generation so linelint doesn't complain. :-)

Initial commit only changed `build.gradle`.  Testing the repro case in the linked issue below also upgraded the minimum gradle version to 7.6, which is a good thing as it is [required for JDK19+ compatibility](https://github.com/gradle/gradle/issues/20372).

### Issues Resolved

Fixes #5494

### Check List
- [x] All tests pass
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dbwiddis** [comment]
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.geo.search.aggregations.bucket.GeoTileGridIT.testGeoShapes
> ```

#6187

File: build.gradle [hunk 1]
```diff
@@ -533,16 +531,12 @@ allprojects {
 wrapper {
   distributionType = 'ALL'
   doLast {
-    final DistributionLocator locator = new DistributionLocator()
```
Comment (reta) [inline build.gradle:536]: @dbwiddis I have a few comments and concerns:
 - the `org.gradle.util.DistributionLocator` seems to be replaced with `org.gradle.util.internal.DistributionLocator` which we should be using
 - by trusting the `distributionUrl` we basically allow anyone to supply own distribution with whatever SHAs it could have, the `DistributionLocator` ensures that SHAs are always coming from the trusted source

Comment (dbwiddis) [reply build.gradle:536]: I would think we should not be using a package marked `internal`.

Note that I sourced the changes from https://github.com/aiven/s3-connector-for-apache-kafka/pull/223 (and confirmed the code).

Comment (reta) [reply build.gradle:536]: > I would think we should not be using a package marked internal.

That would be ideal, I agree, so we could leverage the fact wrapper task has locator which is used in `getDistributionUrl()` method to do exactly the same thing:

```
final URI sha256Uri = new URI(getDistributionUrl() + ".sha256")
``` 

> Note that I sourced the changes from https://github.com/aiven/s3-connector-for-apache-kafka/pull/223 (and confirmed the code).

:-) I would not replicate this approach :)

Comment (dbwiddis) [reply build.gradle:536]: I updated the PR to use `getDistributionUrl` but the locator code is only called if the `distributionUrl` is null, so it doesn't seem to answer your concern...
```java
public String getDistributionUrl() {
    if (distributionUrl != null) {
        return distributionUrl;
    } else if (gradleVersionResolver.getGradleVersion() != null) {
        return locator.getDistributionFor(gradleVersionResolver.getGradleVersion(), distributionType.name().toLowerCase(Locale.ENGLISH)).toString();
    } else {
        return null;
    }
}
```

Comment (dbwiddis) [reply build.gradle:536]: Digging into the code again, I think the `getDistributionUrl()` is an improvement, but I don't see the security concern with specifying a custom URL
 - It looks like an intentional design to allow companies to host their own internal repo
 - the code in the locator (both old and new locations) eventually [breaks down to this](https://github.com/gradle/gradle/blob/b75cbf987d7c9c76b64e274f5997e2769862091b/subprojects/core/src/main/java/org/gradle/util/DistributionLocator.java#LL65C13-L65C127), taking `repoistoryUrl` as an input, which a user can specify.
```java
new URI(repositoryUrl + "/" + archiveName + "-" + version.getVersion() + "-" + archiveClassifier + ".zip");
```
  - All it is really doing is appending a sha256 to the properties file, which a user could do manually anyway if they wanted.  I could just replace the last line here:
```
➜  OpenSearch git:(distribution-locator) cat gradle/wrapper/gradle-wrapper.properties
distributionBase=GRADLE_USER_HOME
distributionPath=wrapper/dists
distributionUrl=https\://services.gradle.org/distributions/gradle-7.6-all.zip
networkTimeout=10000
zipStoreBase=GRADLE_USER_HOME
zipStorePath=wrapper/dists
distributionSha256Sum=312eb12875e1747e05c2f81a4789902d7e4ec5defbd1eefeaccc08acf096505d
```

Comment (reta) [reply build.gradle:536]: May be I am over complicating things :) thanks @dbwiddis

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,17 +1,7 @@
-#
```
Comment (reta) [inline gradle/wrapper/gradle-wrapper.properties:1]: We probably need to keep this header, but since this is auto-generated, may be it is not needed

**Review by reta — APPROVED** [review]

File: gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (dbwiddis) [reply gradle/wrapper/gradle-wrapper.properties:1]: I'm spoiled on SDK where spotless fixes all the license header (and newline) issues for me.  I'll put it back in even if it's not needed, although the next time someone runs `gradle wrapper` it'll delete it again!

**Review by ryanbogan — APPROVED** [review]

**Comment by owaiskazi19** [comment]
```
> Task :qa:mixed-cluster:v2.6.0#mixedClusterTest FAILED
build complete, generating: /var/jenkins/workspace/gradle-check/search/build/10914.tar.bz2

FAILURE: Build failed with an exception.

**Comment by dbwiddis** [comment]
Gradle check failures are the mixed cluster flaky test.

**Comment by reta** [comment]
@dbwiddis could you rebase please? the Gradle's wrapper bits are not needed anymore (we are on 8.0 now), but the dropping `DistributionLocator` is still valid, thank you.

**Comment by dbwiddis** [comment]
> @dbwiddis could you rebase please? the Gradle's wrapper bits are not needed anymore (we are on 8.0 now), but the dropping `DistributionLocator` is still valid, thank you.

Done!

File: buildSrc/src/main/resources/minimumGradleVersion [hunk 3]
```diff
@@ -1 +1 @@
-7.4.1
+8.0
```
Comment (reta) [inline buildSrc/src/main/resources/minimumGradleVersion:1]: @dbwiddis please revert this change, here is the context. We publish a lot Gradle plugins along with OpenSearch core, these plugins are used to run test, publications, and also scaffold OpenSearch plugins. By bumping this minimum version, we basically force everyone to make a switch to Gradle 8, example below for our own `security` plugin:

```
A problem occurred evaluating root project 'security'.
> Failed to apply plugin class 'org.opensearch.gradle.info.GlobalBuildInfoPlugin'.
   > Gradle 8.0+ is required
```

**Review by reta — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/resources/minimumGradleVersion (hunk 3 above)
Comment (dbwiddis) [reply buildSrc/src/main/resources/minimumGradleVersion:1]: ACK.  Do you want it back to 7.4.1 or 7.6 (required for JDK 19+)?

Comment (reta) [reply buildSrc/src/main/resources/minimumGradleVersion:1]: Thanks,  7.4.1 is preferable (we are maintaining this compatibility), gives more freedom to plugin developers.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@dbwiddis backport to 2.x failed :( could you please do it manually? thank you!
